### PR TITLE
Add opcodes for `msg.*`

### PR DESCRIPTION
**Pull request by 0xRampey**
Related to #4 

Adds `msg.value`. `msg.sig` and `msg.data` with light checks in the contract

Also uses enum to pattern match syscalls.

### CODE REVIEW

File: erc20/src/lib.rs [hunk 1]
```diff
@@ -36,6 +39,12 @@ impl ERC20 {
         if owner != address!("0000000000000000000000000000000000000007") {
             revert();
         }
+        if msg_value() != U256::from(0) {
+            revert();
+        }
+        if msg_sig() != [2, 0, 0, 0] {
+            revert();
+        }
```
Comment (gakonst) [inline erc20/src/lib.rs:47]: Should we enforce deterministic 4-byte discriminants here from the function signature, with the ability to override via a `#[sig(0x12345678)` attribute?

File: eth-riscv-runtime/src/lib.rs [hunk 2]
```diff
@@ -116,6 +118,55 @@ pub fn msg_sender() -> Address {
...
 
+pub fn msg_value() -> U256 {
+    let first: u64;
+    let second: u64;
+    let third: u64;
+    let fourth: u64;
+    unsafe {
+        asm!("ecall", lateout("a0") first, lateout("a1") second, lateout("a2") third, lateout("a3") fourth, in("t0") u64::from(Syscall::CALLVALUE));
+    }
+    U256::from_limbs([first, second, third, fourth])
```
Comment (gakonst) [inline eth-riscv-runtime/src/lib.rs:129]: We should add an abstraction for U256 conversion stuff, I think -- not sure best way

**Review by gakonst — DISMISSED** [review]
Conflicts with #9 -- let's get that in first given it touches on the broader syscall interface. Left some thoughts.

File: erc20/src/lib.rs (hunk 1 above)
Comment (leonardoalt) [reply erc20/src/lib.rs:47]: yep makes sense

File: erc20/src/lib.rs [hunk 3]
```diff
@@ -36,6 +39,12 @@ impl ERC20 {
         if owner != address!("0000000000000000000000000000000000000007") {
             revert();
         }
+        if msg_value() != U256::from(0) {
+            revert();
+        }
```
[the remark this answers is not in the document]
Comment (0xRampey) [reply erc20/src/lib.rs:44]: Good idea. Will add this in

File: erc20/src/lib.rs (hunk 1 above)
Comment (0xRampey) [reply erc20/src/lib.rs:47]: Yes, have an issue for it here: https://github.com/r55-eth/r55/issues/7

File: erc20/src/lib.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (0xRampey) [reply erc20/src/lib.rs]: Done @gakonst

File: erc20/src/lib.rs [hunk 4]
```diff
@@ -16,6 +16,9 @@ pub struct ERC20 {
 #[contract]
 impl ERC20 {
     pub fn balance_of(&self, owner: Address) -> u64 {
+        if msg_data()[0] != 0x0 {
```
[the remark this answers is not in the document]
Comment (0xRampey) [reply erc20/src/lib.rs:19]: `msg.data` is not used inside `payable` afaik. 
Do we want to add `receive() external payable{}` to handle no-call-data txs?

File: eth-riscv-runtime/src/lib.rs [hunk 5]
```diff
@@ -41,34 +43,34 @@ use eth_riscv_syscalls::Syscall;
 
 pub fn return_riscv(addr: u64, offset: u64) -> ! {
     unsafe {
-        asm!("ecall", in("a0") addr, in("a1") offset, in("t0") u32::from(Syscall::Return));
+        asm!("ecall", in("a0") addr, in("a1") offset, in("t0") u64::from(Syscall::Return));
```
Comment (leonardoalt) [inline eth-riscv-runtime/src/lib.rs:46]: I think we made these u8 in another PR

File: eth-riscv-runtime/src/lib.rs [hunk 6]
```diff
@@ -116,6 +118,55 @@ pub fn msg_sender() -> Address {
...
+        asm!("ecall", lateout("a0") first, lateout("a1") second, lateout("a2") third, lateout("a3") fourth, in("t0") u64::from(Syscall::CALLVALUE));
+    }
+    U256::from_limbs([first, second, third, fourth])
+}
+
+pub fn msg_sig() -> [u8; 4] {
+    let first: u64;
+    let offset: u64 = 0;
+    unsafe {
+        asm!("ecall", lateout("a0") first, in("a0") offset, in("t0") u64::from(Syscall::CALLDATALOAD));
```
Comment (leonardoalt) [inline eth-riscv-runtime/src/lib.rs:136]: So regarding calldata, currently we dont implement it as a host function, but rather inject it directly as a ROM for the interpreter which lets the contract read "for free". We need to decide which approach we want to use, but I quite like this one.

File: eth-riscv-runtime/src/lib.rs (hunk 5 above)
Comment (0xRampey) [reply eth-riscv-runtime/src/lib.rs:46]: Yup, lemme change them

File: eth-riscv-runtime/src/lib.rs (hunk 6 above)
Comment (0xRampey) [reply eth-riscv-runtime/src/lib.rs:136]: Reading from ROM should be faster than `syscalls`.  Let's change it back to that?

Comment (leonardoalt) [reply eth-riscv-runtime/src/lib.rs]: Yep sounds good

File: erc20/src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (leonardoalt) [reply erc20/src/lib.rs:19]: I think this can be removed now?

File: erc20/src/lib.rs (hunk 1 above)
Comment (leonardoalt) [reply erc20/src/lib.rs:45]: and this too?

File: eth-riscv-syscalls/src/lib.rs [hunk 7]
```diff
@@ -60,4 +60,5 @@ syscalls!(
     (4, Revert, "revert"),
     (5, Caller, "caller"),
     (0x20, Keccak256, "keccak256"),
+    (34, CALLVALUE, "callvalue"),
```
Comment (leonardoalt) [inline eth-riscv-syscalls/src/lib.rs:63]: let's use hex here I think it's more common in EVM world

File: eth-riscv-runtime/src/lib.rs [hunk 8]
```diff
@@ -116,6 +120,29 @@ pub fn msg_sender() -> Address {
...
+pub fn msg_sig() -> [u8; 4] {
+    let sig = unsafe { slice_from_raw_parts(CALLDATA_ADDRESS + 8, 4) };
+    sig.try_into().unwrap()
+}
+
+pub fn msg_data() -> Vec<u8> {
+    let length = unsafe { slice_from_raw_parts(CALLDATA_ADDRESS, 8) };
+    let length = u64::from_le_bytes([length[0], length[1], length[2], length[3], length[4], length[5], length[6], length[7]]) as usize;
+    let calldata = unsafe { slice_from_raw_parts(CALLDATA_ADDRESS + 8, length) };
+    calldata.to_vec()
```
Comment (leonardoalt) [inline eth-riscv-runtime/src/lib.rs:143]: we could return a slice to avoid allocation?

File: contract-derive/src/lib.rs [hunk 9]
```diff
@@ -44,7 +44,15 @@ pub fn contract(_attr: TokenStream, item: TokenStream) -> TokenStream {
         }).collect();
 
         let arg_names: Vec<_> = (0..method.sig.inputs.len() - 1).map(|i| format_ident!("arg{}", i)).collect();
-
+        let checks = if !is_payable(&method) {
```
Comment (leonardoalt) [inline contract-derive/src/lib.rs:47]: Nice!

File: eth-riscv-runtime/src/lib.rs (hunk 8 above)
Comment (0xRampey) [reply eth-riscv-runtime/src/lib.rs:143]: Ah yes!

File: erc20/src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (0xRampey) [reply erc20/src/lib.rs:19]: Yeah sounds good.

**Comment by 0xRampey** [comment]
Not much conflict with #9 now. I think we can proceed to merge this?

**Review by leonardoalt — APPROVED** [review]
Awesome!

**Pull request merged**